x

```diff
@@ -15,19 +21,65 @@ realpath() {
     fi
 }
 
-xpy=$(dirname "$(realpath "$0")")/x.py
+exists() {
+    command -v "$1" >/dev/null 2>&1
+}
+
+run_python() {
+    xpy=$(dirname "$(realpath "$0")")/x.py
 
-# On Windows, `py -3` sometimes works. We need to try it first because `python3`
-# sometimes tries to launch the app store on Windows.
-for SEARCH_PYTHON in py python3 python python2; do
-    if python=$(command -v $SEARCH_PYTHON) && [ -x "$python" ]; then
-        if [ $SEARCH_PYTHON = py ]; then
-            extra_arg="-3"
-        else
-            extra_arg=""
+    # Modern Linux and macOS systems commonly only have a thing called `python3` and
+    # not `python`, while Windows commonly does not have `python3`, so we cannot
+    # directly use python in the x.py shebang and have it consistently work. Instead we
+    # have a shell script to look for a python to run x.py.
+    #
+    # On Windows, `py -3` sometimes works. We need to try it first because `python3`
+    # sometimes tries to launch the app store on Windows.
+    for SEARCH_PYTHON in py python3 python python2; do
+        if python=$(command -v $SEARCH_PYTHON) && [ -x "$python" ]; then
+            if [ $SEARCH_PYTHON = py ]; then
+                extra_arg="-3"
+            else
+                extra_arg=""
+            fi
+            exec "$python" $extra_arg "$xpy" "$@"
         fi
-        exec "$python" $extra_arg "$xpy" "$@"
+    done
+    echo "$0: error: did not find python installed" >&2
+    exit 1
+}
+
+run_downloaded_binary() {
+    # For now, we just hard-code `./build` for the bootstrap binary.
+    # Bootstrap itself will still respect `--build-dir`.
+    cache_dir=build/cache/bootstrap-$latest_commit
+    mkdir -p "$cache_dir"
+
+    platform=$(rustc --version --verbose | grep host | cut -d ' ' -f2)
+
+    tarfile=bootstrap-nightly-$platform.tar.xz
+    if ! [ -e "$cache_dir/$tarfile" ]; then
+        download https://ci-artifacts.rust-lang.org/rustc-builds/4652f5eb25b76c24fc40b7267d87a07d5876901f/$tarfile "$cache_dir/$tarfile"
+    fi
+
+    if ! [ -e build/bootstrap/bin ]; then
+        rm -rf build/bootstrap  # make sure we extract this to the top-level, not as a subdirectory of the target dir
+        mkdir -p build/bootstrap
+        tar -C build -xf "$cache_dir/$tarfile" bootstrap-nightly-$platform/bootstrap/bootstrap --strip-components=2
     fi
-done
-echo "$0: error: did not find python installed" >&2
-exit 1
+
+    exec build/bootstrap/bin/bootstrap "$@"
+}
+
+if ! exists rustc; then
+    # We don't support platform detection based on `uname`, etc; make python do it.
+    run_python "$@"
+fi
+
+latest_commit=$(git rev-list HEAD --author=bors@rust-lang.org --max-count 1)
+if git diff-index --quiet $latest_commit -- src/bootstrap; then
```

Comment:
Should this be conditional on a config.toml config? A distro probably doesn't want to trigger this logic even if according to git they didn't change bootstrap.